### PR TITLE
Optimize Enum.with_index/2 for lists

### PR DESCRIPTION
**Pull request by sabiwara**
Hi,

I noticed that `Enum.with_index/2` could be optimized for lists using a specialized implementation
- x2 on Intel Chip, see [bench](https://github.com/sabiwara/elixir_benches/blob/main/bench/fast_enum_with_index.results.txt)
- x3 on M1 Chip / OTP 25, see [bench](https://github.com/sabiwara/elixir_benches/blob/main/bench/fast_enum_with_index.results_m1.txt)

~I tried both tail-recursion and [body-recursion](https://github.com/sabiwara/elixir_benches/blob/main/lib/fast_enum.ex#L438-L442=), and went with tail-recursion because it was considerably faster, especially on ARM (x2).~

Edit: after the additional benchmarks from below, I updated to body-recursion which consistently uses less memory. Let me know your thoughts.

### CODE REVIEW

**Comment by whatyouhide** [comment]
Did you try this with multiple list sizes, to compare tail recursion and body recursion (both execution time and memory usage)? In general I like the optimization over the generic enum implementation, just curious about the two possible implementations of the optimization 😉

**Review by whatyouhide — APPROVED** [review]

**Comment by sabiwara** [comment]
> Did you try this with multiple list sizes, to compare tail recursion and body recursion (both execution time and memory usage)? In general I like the optimization over the generic enum implementation, just curious about the two possible implementations of the optimization 😉

Thanks for the advice, I totally forgot to check these. This gave very [interesting results](https://github.com/sabiwara/elixir_benches/blob/main/bench/fast_enum_with_index_inputs.results.txt) which might make the body-recursion more interesting overall.

TLDR:
- tail recursion uses ~30% more memory
- tail recursion is yet much faster from 100~10_000
- but for small (10) or big (100_000) lists the body recursion wins again (although for small ones I'm getting fast execution warnings so might not be so relevant)

Maybe optimizing for the memory seems like a safer bet here, happy to re-push using body-recursion 🙂

**Comment by whatyouhide** [comment]
I want to hear from @josevalim too, but I would probably optimize for speed in the sub-1000 range. Most of the time I use `with_index` is with smaller lists (in general, smaller lists are more common in my experience). For those, the speed can accumulate. Memory too, but the memory usage difference is small in absolute terms. It's a compromise anyways, so whatever we pick is gonna be okay in my opinion 😉

**Comment by josevalim** [comment]
The general guidelines is that we should not actually prioritize those numbers as they may change any time or even in different systems. We should consider both styles as largely equivalent, except in very few cases, so go with whatever is more readable. More info: https://www.erlang.org/doc/efficiency_guide/myths.html#myth--tail-recursive-functions-are-much-faster-----than-recursive-functions

**Review by josevalim — APPROVED** [review]

**Comment by whatyouhide** [comment]
Thanks @sabiwara! 💟
